### PR TITLE
Core - Add support for storage deposit calculations

### LINKED ISSUE #337: Core - Add support for storage deposit calculations
## Description

Storage deposit should be calculated when adding a new output to the outputs list.

Weights for different outputs are listed here:
https://github.com/muXxer/protocol-rfcs/blob/master/tips/TIP-0019/tip-0019.md#outputs

Referenced implementation in Go:
https://github.com/iotaledger/iota.go/blob/v3/output.go#L663

A storage deposit calculator implemented in Python if needed:
https://github.com/iotaledger/dust-protection-calculations

### PR DESCRIPTION
**Pull request by matjazv**
# Description of change

Fixes #337 

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

New unit tests are in `core/test_storage_deposit.c`

## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests using CTest that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/core/models/outputs/byte_cost_config.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
+// Copyright 2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef __CORE_BYTE_COST_CONFIG_H__
+#define __CORE_BYTE_COST_CONFIG_H__
```
Comment (oopsmonk) [inline src/core/models/outputs/byte_cost_config.h]:
```suggestion
#ifndef __CORE_MODELS_OUTPUTS_BYTE_COST_CONFIG_H__
#define __CORE_MODELS_OUTPUTS_BYTE_COST_CONFIG_H__
```

File: src/core/models/outputs/byte_cost_config.h [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ *
+ * @param[in] config A byte cost configuration
+ */
+void byte_cost_config_free(byte_cost_config_t *config);
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif  // __CORE_BYTE_COST_CONFIG_H__
```
Comment (oopsmonk) [inline src/core/models/outputs/byte_cost_config.h]:
```suggestion
#endif  // __CORE_MODELS_OUTPUTS_BYTE_COST_CONFIG_H__
```

File: src/core/models/outputs/storage_deposit.c [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+// Notice, this solution is a trade-off for memory optimization that we don't create the basic output and calculate byte
+// cost from it.
+static uint64_t basic_address_storage_deposit(byte_cost_config_t *config, address_t *addr) {
+  if (config == NULL || addr == NULL) {
+    printf("[%s:%d] invalid parameters\n", __func__, __LINE__);
+    return UINT64_MAX;
+  }
+
+  // output serialized length = output type + amount + native tokens + unlock count + block count
+  uint64_t output_serialized_len = 12;  // 1 + 8 + 1 + 1
```
Comment (oopsmonk) [inline src/core/models/outputs/storage_deposit.c]:
```suggestion
  // output serialized length = output type + amount + native tokens + unlock count + block count
  uint64_t output_serialized_len = 12;  // 1 + 8 + 1 + 1 + 1
```

File: src/core/models/outputs/storage_deposit.h [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+/**
+ * @brief Check if a sufficient storage deposit was made for the given output
+ *
+ * @param[in] config A byte cost configuration
+ * @param[in] output_type UTXO output type
+ * @param[in] output Pointer to an output
+ * @return true if output has enough storage deposit amount
+ */
+bool storage_deposit_sufficient_output_deposit_check(byte_cost_config_t *config, utxo_output_type_t output_type,
```
Comment (oopsmonk) [inline src/core/models/outputs/storage_deposit.h]: we know it checks if the output has sufficient deposit, therefore `sufficient output_deposit` is redundant, the function name can be shorter.
```suggestion
bool storage_deposit_check(byte_cost_config_t *config, utxo_output_type_t output_type,
```

**Review by oopsmonk — CHANGES_REQUESTED** [review]

File: src/core/models/outputs/byte_cost_config.h (hunk 1 above)
Comment (matjazv) [reply src/core/models/outputs/byte_cost_config.h]: Done.

File: src/core/models/outputs/byte_cost_config.h (hunk 2 above)
Comment (matjazv) [reply src/core/models/outputs/byte_cost_config.h]: Done.

File: src/core/models/outputs/storage_deposit.h (hunk 4 above)
Comment (matjazv) [reply src/core/models/outputs/storage_deposit.h]: Done. Thanks.

File: src/core/models/outputs/storage_deposit.c (hunk 3 above)
Comment (matjazv) [reply src/core/models/outputs/storage_deposit.c]: Done.

**Review by oopsmonk — APPROVED** [review]

**Pull request merged**